File: frontend/src/services/commerceService.js
import { apiRequest } from "./apiClient";

function unwrap(response) {
  return response?.data;
}

function toQuery(params = {}) {
  const query = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null && value !== "") {
      query.set(key, value);
    }
  });
  return query.toString();
}

export async function getCart() {
  return unwrap(await apiRequest("/cart"));
}

export async function addCartItem(courseId) {
  return unwrap(await apiRequest("/cart/items", {
    method: "POST",
    body: JSON.stringify({ courseId }),
  }));
}

export async function removeCartItem(courseId) {
  return unwrap(await apiRequest(`/cart/items/${courseId}`, { method: "DELETE" }));
}

export async function clearCart() {
  return unwrap(await apiRequest("/cart", { method: "DELETE" }));
}

export async function applyCoupon(code) {
  return unwrap(await apiRequest("/cart/coupon", {
    method: "POST",
    body: JSON.stringify({ code }),
  }));
}

export async function removeCoupon() {
  return unwrap(await apiRequest("/cart/coupon", { method: "DELETE" }));
}

export async function createOrder() {
  return unwrap(await apiRequest("/orders", { method: "POST" }));
}

export async function getOrders(params = {}) {
  const query = toQuery(params);
  return unwrap(await apiRequest(`/orders${query ? `?${query}` : ""}`));
}

export async function getOrder(orderCode) {
  return unwrap(await apiRequest(`/orders/${orderCode}`));
}

export async function cancelOrder(orderCode) {
  return unwrap(await apiRequest(`/orders/${orderCode}/cancel`, { method: "POST" }));
}

export async function createPayment(orderCode) {
  return unwrap(await apiRequest(`/payments/${orderCode}/create`, { method: "POST" }));
}

export async function getPaymentStatus(orderCode) {
  return unwrap(await apiRequest(`/payments/${orderCode}/status`));
}

export async function getAdminOrders(params = {}) {
  const query = toQuery(params);
  return unwrap(await apiRequest(`/admin/orders${query ? `?${query}` : ""}`));
}

export async function getAdminTransactions(params = {}) {
  const query = toQuery(params);
  return unwrap(await apiRequest(`/admin/transactions${query ? `?${query}` : ""}`));
}

export async function getAdminCoupons(params = {}) {
  const query = toQuery(params);
  return unwrap(await apiRequest(`/admin/coupons${query ? `?${query}` : ""}`));
}

export async function getAdminRefunds(params = {}) {
  const query = toQuery(params);
  return unwrap(await apiRequest(`/admin/refunds${query ? `?${query}` : ""}`));
}

export async function activateCoupon(id) {
  return unwrap(await apiRequest(`/admin/coupons/${id}/activate`, { method: "POST" }));
}

export async function deactivateCoupon(id) {
  return unwrap(await apiRequest(`/admin/coupons/${id}/deactivate`, { method: "POST" }));
}

export async function approveRefund(id, note = "") {
  return unwrap(await apiRequest(`/admin/refunds/${id}/approve`, {
    method: "POST",
    body: JSON.stringify({ note }),
  }));
}

export async function rejectRefund(id, note = "") {
  return unwrap(await apiRequest(`/admin/refunds/${id}/reject`, {
    method: "POST",
    body: JSON.stringify({ note }),
  }));
}

export async function getAdminDashboardStats() {
  return unwrap(await apiRequest("/admin/dashboard/stats"));
}

File: backend/src/main/java/com/example/englishlearning/service/CommerceService.java
package com.example.englishlearning.service;

import com.example.englishlearning.dto.common.PageResponse;
import com.example.englishlearning.dto.payment.CartResponse;
import com.example.englishlearning.dto.payment.CouponResponse;
import com.example.englishlearning.dto.payment.OrderResponse;
import com.example.englishlearning.dto.payment.PaymentResponse;
import com.example.englishlearning.dto.payment.PaymentWebhookRequest;
import com.example.englishlearning.dto.payment.RefundResponse;
import com.example.englishlearning.dto.payment.TransactionResponse;
import org.springframework.data.domain.Pageable;

import java.util.Map;

public interface CommerceService {

    CartResponse getCart(String email);

    CartResponse addCartItem(String email, Long courseId);

    CartResponse removeCartItem(String email, Long courseId);

    CartResponse clearCart(String email);

    CartResponse applyCoupon(String email, String code);

    CartResponse removeCoupon(String email);

    OrderResponse createOrderFromCart(String email);

    PageResponse<OrderResponse> getMyOrders(String email, Pageable pageable);

    OrderResponse getMyOrder(String email, String orderCode);

    OrderResponse cancelMyOrder(String email, String orderCode);

    PaymentResponse createPayment(String email, String orderCode);

    PaymentResponse handlePaymentReturn(Map<String, String> params);

    PaymentResponse handlePaymentWebhook(PaymentWebhookRequest request);

    PaymentResponse getPaymentStatus(String email, String orderCode);

    PageResponse<OrderResponse> getAdminOrders(Pageable pageable);

    PageResponse<TransactionResponse> getAdminTransactions(Pageable pageable);

    PageResponse<CouponResponse> getAdminCoupons(Pageable pageable);

    PageResponse<RefundResponse> getAdminRefunds(Pageable pageable);

    CouponResponse activateCoupon(Long id);

    CouponResponse deactivateCoupon(Long id);

    RefundResponse approveRefund(Long id, String note);

    RefundResponse rejectRefund(Long id, String note);
}
